### PR TITLE
Force return types and arrow callbacks on functions

### PR DESCRIPTION
**Pull request by ferferga**
Added a rule for enforcing return types on all the functions, which should help us with type checking across all the code

We should also investigate how we can get ``VueI18n`` to return strings instead of ``TranslateResult``, so we don't need to invoke ``toString()`` or do any string casting.

### CODE REVIEW

**Comment by ThibaultNocchi** [comment]
Maybe a simple global mixin to wrap the calls to i18n?

**Comment by ferferga** [comment]
@ThibaultNocchi Just found this: https://github.com/kazupon/vue-i18n/issues/410#issuecomment-538465940

Making a commit to replace all occurences

File: components/Buttons/FilterButton.vue [hunk 1]
```diff
@@ -9,7 +9,7 @@
         v-bind="attrs"
         v-on="on"
       >
-        {{ $t('filter') }}
+        {{ $tc('filter') }}
```
Comment (MrTimscampi) [inline components/Buttons/FilterButton.vue:12]: I'm not in favor of this.

`$tc()` is meant to handle plurals and should only be used to do that, imo.

I don't see the issue with using `.toString()` in TypeScript code. Templates won't need it and it doesn't add much.

Overall, this feels very hacky and like an abuse of the vue-i18n APIs.

I wouldn't be in favor of wrapping `$t()` either, mind you (We should avoid wrapping things up for the hell of it).

**Comment by ThibaultNocchi** [comment]
@ferferga In the Vue i18n API, `$tc` also returns a `TranslateResult`, so I guess your typings should work with `$t`

File: components/Buttons/FilterButton.vue (hunk 1 above)
Comment (ferferga) [reply components/Buttons/FilterButton.vue:12]: @MrTimscampi What about extending the interface?

```
declare module 'vue/types/vue' {
interface Vue {
$t: (name: string, attrs?: any) => string
```

**Comment by ferferga** [comment]
@ThibaultNocchi That's what they say but doesn't translate into practice 🤷‍♂️

File: components/Buttons/FilterButton.vue (hunk 1 above)
Comment (MrTimscampi) [reply components/Buttons/FilterButton.vue:12]: That should be fine for our use. It's a string or a set of key/values, but we should always get the string version with how we use it.

See: https://github.com/kazupon/vue-i18n/blob/v8.x/types/index.d.ts#L18

File: layouts/default.vue [hunk 2]
```diff
@@ -143,7 +143,7 @@ export default Vue.extend({
           };
         })
     }),
-    items() {
+    items(): Array<Record<string, string>> {
```
Comment (Camc314) [inline layouts/default.vue:146]: This should be a custom type

File: layouts/default.vue [hunk 3]
```diff
@@ -152,7 +152,7 @@ export default Vue.extend({
         }
       ];
     },
-    configItems() {
+    configItems(): Array<Record<string, string>> {
```
Comment (Camc314) [inline layouts/default.vue:155]: This should be a custom type

File: pages/person/_itemId/index.vue [hunk 4]
```diff
@@ -172,7 +172,7 @@ export default Vue.extend({
       }
     },
     birthPlace: {
-      get() {
+      get(): null {
```
Comment (Camc314) [inline pages/person/_itemId/index.vue:175]:
```suggestion
      get(): string | null {

File: pages/person/_itemId/index.vue [hunk 5]
```diff
@@ -190,7 +190,7 @@ export default Vue.extend({
       }
     },
     shows: {
-      get() {
+      get(): Array<Record<string, string>> {
```
Comment (Camc314) [inline pages/person/_itemId/index.vue:193]:
```suggestion
      get(): BaseItemDto[] {

File: plugins/browserDetection.ts [hunk 6]
```diff
@@ -100,10 +100,10 @@ class BrowserDetector {
...
    */
-  isApple() {
-    return (
-      navigator.vendor && navigator.vendor.includes('Apple') && !this.isTizen()
-    );
+  isApple(): boolean {
+    return (navigator.vendor &&
+      navigator.vendor.includes('Apple') &&
+      !this.isTizen()) as boolean;
   }
```
Comment (Camc314) [inline plugins/browserDetection.ts:107]:
```suggestion
  return navigator?.vendor.includes('Apple') && !this.isTizen();

File: store/homeSection.ts [hunk 7]
```diff
@@ -45,7 +45,9 @@ type MutationPayload = {
 };
 
 export const getters: GetterTree<HomeSectionState, AppState> = {
-  getHomeSectionContent: (state) => (section: HomeSection) => {
+  getHomeSectionContent: (state) => (
+    section: HomeSection
+  ): Array<never> | BaseItemDto | BaseItemDto[] => {
```
Comment (Camc314) [inline store/homeSection.ts:50]:
```suggestion
  ): BaseItemDto[] => {

File: store/tvShows.ts [hunk 8]
```diff
@@ -31,10 +31,14 @@ type MutationPayload = {
...
-  getSeasons: (state) => ({ itemId }: { itemId: string }) => {
+  getSeasons: (state) => ({ itemId }: { itemId: string }): BaseItemDto[] => {
     return state[itemId]?.seasons;
   },
-  getSeasonEpisodes: (state) => ({ itemId }: { itemId: string }) => {
+  getSeasonEpisodes: (state) => ({
+    itemId
+  }: {
+    itemId: string;
+  }): Array<BaseItemDto[]> => {
```
Comment (Camc314) [inline store/tvShows.ts:41]:
```suggestion
  }): BaseItemDto[][] => {

File: utils/hlsFormats.ts [hunk 9]
```diff
@@ -7,7 +7,9 @@ import { browserDetector } from '~/plugins/browserDetection';
  * @param {HTMLVideoElement} videoTestElement A HTML video element for testing codecs
  * @returns {boolean} Determines if the browser has AC3 in HLS support
  */
-function supportsAc3InHls(videoTestElement: HTMLVideoElement) {
+function supportsAc3InHls(
+  videoTestElement: HTMLVideoElement
+): boolean | string {
```
Comment (Camc314) [inline utils/hlsFormats.ts:12]:
```suggestion
): boolean {

File: utils/mp4AudioFormats.ts [hunk 10]
```diff
@@ -57,7 +57,7 @@ function hasMp2AudioSupport(): boolean {
  * @param {HTMLVideoElement} videoTestElement A HTML video element for testing codecs
  * @returns {boolean} Determines if browserr has DTS audio support
  */
-function hasDtsSupport(videoTestElement: HTMLVideoElement) {
+function hasDtsSupport(videoTestElement: HTMLVideoElement): boolean | string {
```
Comment (Camc314) [inline utils/mp4AudioFormats.ts:60]:
```suggestion
function hasDtsSupport(videoTestElement: HTMLVideoElement): boolean {

File: components/Forms/AddServerForm.vue [hunk 11]
```diff
@@ -57,7 +57,7 @@ export default Vue.extend({
   },
   methods: {
     ...mapActions('servers', ['connectServer']),
-    connectToServer() {
+    connectToServer(): void {
```
[the remark this answers is not in the document]
Comment (ferferga) [reply components/Forms/AddServerForm.vue:60]: @Camc314 ``connectServer`` (from store) is ``async``, this one isn't.

File: components/Item/Metadata/PersonEditor.vue [hunk 12]
```diff
@@ -65,7 +65,7 @@ export default Vue.extend({
   props: {
     person: {
       type: Object,
-      default: () => ({
+      default: (): Record<string, string> => ({
```
[the remark this answers is not in the document]
Comment (ferferga) [reply components/Item/Metadata/PersonEditor.vue:68]: @Camc314 I tracked this down and parent item is passing ``BaseItemPerson``, so I used it

File: layouts/default.vue (hunk 2 above)
Comment (ferferga) [reply layouts/default.vue:146]: @Camc314 Created ``LayoutButton`` type for this.

File: layouts/default.vue (hunk 3 above)
Comment (ferferga) [reply layouts/default.vue:155]: @Camc314 Created ``LayoutButton`` type for this.

File: pages/person/_itemId/index.vue (hunk 4 above)
Comment (ferferga) [reply pages/person/_itemId/index.vue:175]: @Camc314 Good catch

File: utils/hlsFormats.ts (hunk 9 above)
Comment (ferferga) [reply utils/hlsFormats.ts:12]: @Camc314 This function also returns the string from the ``canPlayType`` https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/canPlayType

File: utils/mp4AudioFormats.ts (hunk 10 above)
Comment (ferferga) [reply utils/mp4AudioFormats.ts:60]: @Camc314 Same as above

File: components/Buttons/UserButton.vue [hunk 13]
```diff
@@ -36,11 +36,11 @@ export default Vue.extend({
...
-    menuItems() {
+    menuItems(): Array<Record<never, never>> {
       return [
         {
           title: this.$t('logout'),
-          action: () => {
+          action: (): void => {
             this.logoutUser();
           }
         }
```
[the remark this answers is not in the document]
Comment (ferferga) [reply components/Buttons/UserButton.vue:46]: @Camc314 Created ``MenuItem`` type for this

**Review by MrTimscampi — APPROVED** [review]
LGTM

File: components/Forms/AddServerForm.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (Camc314) [reply components/Forms/AddServerForm.vue:60]: Then there should be an `await` on `this.connectServer(this.serverUrl);`

[the remark this answers is not in the document]
Comment (ferferga) [reply components/Forms/AddServerForm.vue:60]: @Camc314 Fixed on #482, not sure why this wasn't async in the first place

**Pull request merged**